settings/settings.md
# Settings file

## Description

These settings file control the simulation and provide a large number of customization options. All options have defaults which will be used if the setting is not found and can be overridden by the user, these are indicated below by the value after the parameter name. The settings file is a JSON file and can be edited with any text editor. Some settings have a CLI option which will take precedence over the file settings. These settings can be placed in any order in the file.

## Settings

- `cosmo_params` : Dictionary of cosmological parameters used by CAMB. This will be passed into CAMB via the [camb.set_params](https://camb.readthedocs.io/en/latest/results.html#camb.results.CAMBdata.set_params) function, and supports any CAMB option supported there. The only required options have the following defaults:
  - `As`, 2.13e-09: Amplitude of the primordial power spectrum.
  - `pivot_scalar`, 0.05 : Pivot scale for the primordial power spectrum.
  - `ns`, 0.9624 : Spectral index of the primordial power spectrum.
  - `lmax`: Will be set by the `lmax` value below, do not set it here as it will be overridden.

- `seed` : Seed for the random number generator. This is used to generate the random noise and the random field. The code has not been fully tested for reproducibility but this should help. There is no default as the default is to use a random seed.
- `force_generation`, False : Boolean to force the generation of the CMB maps. If this is set to true the code will generate the CMB maps even if they already exist, otherwise it will check for existing data files and stop running. The code has been designed so that if a run partially fails you can restart it and it will only generate the missing data, this will override that behavior.
- `lensing`, True : Boolean to include lensing in the simulation. If true the lensing potential will be generated and added to the CMB map.
- `pols`, `T` : a string of [T, E, TE] for which polarization in the simulation.
- `double_precision`, False : Boolean to use double precision. This will increase the memory usage and has not been tested much.

- `nsims`, 100 : Number of simulations to generate per run.
- `ndups`, 25 : Number of sims to make per A_lm calculation, giving multiple sims created from the same background.
- `narray`, 1 : Number of runs that will be used. This allows using slurm to run multiple simulations in parallel while limiting the resources used. This does not actually run the simulations so you will need to match this with slurm settings.
- `nside`, 128 : Healpix resolution parameter. This is the resolution of the Healpix map used in the simulation.

  - > Note: The code used healpy's pixel weights to improve accuracy, and thus only supports a limited number of nsides found [here](https://github.com/healpy/healpy-data/tree/master/full_weights). Currently these are [32, 64, 128, 256, 512, 1024, 2048, 4096]. To change this behavior set `use_pixel_weights=False` in the map2alm function calls.

- `mc_steps`, 300 : Number of Monte Carlo steps used to initialize the KSW estimator.
- `estimate`, True : Boolean to compute KSW estimates during generation.
- `num_estimates`, 1000 : Number of estimates to use in the KSW estimator. This is useful if you are using a large number of simulations and want to use a subset of them for time.
- `save_ksw`, False : Boolean to save the KSW MC states to disk.
- `shapes`, `all` : Bispectrum shapes to generate. One of `local`, `equilateral`, `orthogonal`, or `all`, or a list of them.
- `phi_scale`, 1 : Scaling factor applied to the lensing potential.
- `lmin`, 2 : The lmin parameter used in calculations.
- `lmax`, 3*nside - 1 : The lmax parameter used in calculations. Defaults to optimal value for the nside parameter based on the [anafast documentation](https://healpix.sourceforge.io/html/fac_anafast.htm).
- `lmax_buffer`, 128 : Buffer for the lmax parameter used in some camb calculations to ensure that the code can calculate the required power spectra and lensing to sufficient lmax.
- `fnl_range`, [-1000, 1000] : Range of fNL values to use in the KSW estimator.

- `noise`, True : Boolean to include noise and beaming in the simulation.
- `beam_width`, 1 : Beam width in arcmin.
- `noise_tt`, 1 : Noise scale for the TT power spectrum, in arcmin.
- `noise_ee`, 5 : Noise scale for the EE power spectrum, in arcmin.

- `r_min`, 0 : The minimum value of r to use in the simulation.
- `r_max`, 50000 : The maximum value of r to use in the simulation.
  - > See [Smith and Zaldarriaga 2011](https://arxiv.org/abs/astro-ph/0612571) table 2 for information on the r values used in the simulation.

- `base_name` : Base name for the output files. This will be used to generate the output file names and is automatically generated based on settings if not provided.
- `base_dir`, "data" : Base directory for the output files. It is recommend to put this on a performative file system.
- `plot_dir`, "plots" : Directory for the output plots. This should be a subdirectory of `base_dir`.
- `model_dir`, "models" : Directory for the output models. This should be a subdirectory of `base_dir`.
- `data_dir`, "data" : Directory for the output data. This should be a subdirectory of `base_dir`.
- `mc_dir`, "kswmc" : Directory for the output monte carlo data. This should be a subdirectory of `base_dir`.
- `tb_dir`, "tensorboard" : Directory for TensorBoard logs. This should be a subdirectory of `base_dir`.
- `wandb_dir`, "wandb" : Directory for Weights & Biases logs. This should be a subdirectory of `base_dir`.

- `plot`, True : Boolean to generate plots. If false the code will not generate plots.
- `save_alms`, False : Boolean to save the alm data. If false the code will not save the alm data.
- `tensorboard`, False : Boolean to enable TensorBoard logging for the trainer.
- `wandb`, False : Boolean to enable Weights & Biases logging for the trainer.
- `tf_cache`, True : Boolean to enable file-based disk caching of the TensorFlow datasets.
- `tf_mem_cache`, True : Boolean to enable in-memory caching of the TensorFlow datasets.

## CLI args

The scripts can be run with a number of command line arguments which will override the settings file. This is to help with scripted runs. These are:

- `--nsims`, int : Number of simulations to generate per run.
- `--narray`, int : Number of runs that will be used.
- `--nside`, int : Healpix resolution parameter.
- `--ndups`, int : Number of sims to make per A_lm calculation, giving multiple sims created from the same background.
- `--base_dir`, str : Base directory for the output files.
- `--seed`, int : Seed for the random number generator.
- `--base_name`, str : Base name for the output files.
- `--lmin`, int : The lmin parameter used.
- `--lmax`, int : The lmax parameter used.
- `--lmax_buffer`, int : Buffer for the lmax parameter used in some camb calculations to ensure that the code can calculate the required power spectra to sufficient lmax.
- `--fnl_range`, [int, int] : Range of fNL values to use in the KSW estimator.
- `--num_estimates`, int : Number of estimates to use in the KSW estimator.
- `--mc_steps`, int : Number of Monte Carlo steps used to initialize the KSW estimator.
- `--phi_scale`, float : Scaling factor for the lensing potential.
- `--shapes`, [str, ...] : Bispectrum shapes to generate: local, equilateral, orthogonal, or all.
- `--pols`, str : a string of [T, E] for which polarization in the simulation.

The following arguments can be provided as is or with a `--no` prefix to negate them, e.g. `--no-lensing`:

- `--lensing` : Force lensing in the simulation.
- `--noise` : Force noise and beaming in the simulation.
- `--force_generation` : Force the generation of the CMB maps, overriding existing data.
- `--estimate` : Compute KSW estimates during generation.
- `--double_precision` : Force double precision.
- `--plot` : Force generation of plots.
- `--save_alms` : Save the combined alms.
- `--save_ksw` : Save the KSW MC states.
- `--tensorboard` : Enable TensorBoard logging.
- `--wandb` : Enable Weights & Biases logging.
- `--tf_cache` : Enable disk caching of TensorFlow datasets.
- `--tf_mem_cache` : Enable in-memory caching of TensorFlow datasets.

If scripting runs you can save the settings with:

- `--save_settings` : Save the settings to a file, including CLI changes. This will save to `base_dir/settings/<job>_<name>.json`.

> Note: the save_settings option is the only option which must be provided by CLI, otherwise it would do nothing but make a copy.

## Full example with defaults

```JSON
{
    "cosmo_params" : {
        "As": 2.13e-09,
        "pivot_scalar": 0.05,
        "ns": 0.9624
    },
    "seed": null,
    "force_generation": false,
    "lensing": true,
    "pols": "T",
    "double_precision": false,

    "nsims" : 100,
    "ndups": 25,
    "narray": 1,
    "nside": 128,

    "mc_steps": 300,
    "estimate": true,
    "num_estimates": 1000,
    "save_ksw": false,
    "shapes": "all",
    "phi_scale": 1,
    "lmin": 2,
    "lmax": 383,
    "lmax_buffer": 128,
    "fnl_range": [-1000, 1000],

    "noise": true,
    "beam_width" : 1,
    "noise_tt" : 1,
    "noise_ee" : 5,

    "r_min": 0,
    "r_max": 50000,

    "plot": true,
    "save_alms": false,
    "tensorboard": false,
    "wandb": false,
    "tf_cache": true,
    "tf_mem_cache": true,

    "base_name": null,
    "base_dir": "data",
    "plot_dir": "plots",
    "model_dir": "models",
    "data_dir": "data",
    "mc_dir": "kswmc",
    "tb_dir": "tensorboard",
    "wandb_dir": "wandb"
}
```

> For more information look at the `mlpng/core.py` file.


scripts/utils/config.py
import json
import os
import argparse

import healpy as hp
import numpy as np
from astropy import units as u

import matplotlib.pyplot as plt

import logging

logger = logging.getLogger(__name__)


# this could be cleaned up and improved a lot but little reward
def parse_args(args):
    # note: arguments must be stored in a var name matching the dict key
    parser = argparse.ArgumentParser()
    parser.add_argument("settings_file")
    parser.add_argument("--nsims", type=int)
    parser.add_argument("--narray", type=int)
    parser.add_argument("--npatches", type=int)
    parser.add_argument("--lensing", action=argparse.BooleanOptionalAction)
    parser.add_argument("--noise", action=argparse.BooleanOptionalAction)
    parser.add_argument("--force_alm_gen", action=argparse.BooleanOptionalAction)
    parser.add_argument("--fnl_range", type=float, nargs=2)
    parser.add_argument("--base_dir", type=str)
    parser.add_argument("--save_settings", action="store_true")
    parser.add_argument("--seed", type=int)
    parser.add_argument("--base_name", type=str)
    parser.add_argument("--noise_scale_tt", type=float)
    parser.add_argument("--beam_width", type=float)

    logger.debug(f"parsing cli args: {args}")
    return parser.parse_args(args)


def get_cosmo_defaults():
    defaults = {
        "As": 2.13e-09,
        "ns": 0.9624,
        "pivot_scalar": 0.05,
        "max_l": 1000,
        "lmax": 500,
    }
    return defaults


class Config:
    def __init__(self, argv=None):
        args = parse_args(argv)

        logger.info(f"Loading settings from file {args.settings_file}")
        with open(args.settings_file, "r") as f:
            self.settings = json.load(f)

        # replace the settings with the command line arguments
        for key, value in vars(args).items():
            if value is not None and key not in ["settings_file", "save_settings"]:
                logger.debug(f"Forcing setting {key} to {value}")
                self.settings[key] = value

        # set the cosmological parameters defaults
        self.settings["cosmo_params"] = {
            **get_cosmo_defaults(),
            **self.settings.get("cosmo_params", {}),
        }
        if logger.getEffectiveLevel() <= logging.DEBUG:
            # Log the differences
            defaults = get_cosmo_defaults()
            overrides = self.settings.get("cosmo_params", {})
            for key in defaults.keys():
                if key in overrides and defaults[key] != overrides[key]:
                    logger.debug(
                        f"Overriding cosmo param {key} from {defaults[key]} to {overrides[key]}"
                    )

        # just clean up the code below a little
        settings = self.settings
        self.cosmo_params = self.settings["cosmo_params"]

        logger.info(f"Running with settings: \n{json.dumps(settings, indent=2)}")

        # main parameters
        self.lmax = self.cosmo_params["lmax"]
        self.nside = settings.get("nside", 1024)
        self.patch_side_deg = settings.get("patch_side_deg", 10)
        self.pols = settings.get("polarizations", "T")
        self.npol = len(self.pols)
        self.fnl_min, self.fnl_max = settings.get("fnl_range", (-1, 1))
        self.lensing = settings.get("lensing", False)
        self.force_alm_gen = settings.get("force_alm_gen", False)

        self.seed = settings.get("seed", np.random.default_rng().integers(0, 2**32 - 1))
        self.rng = np.random.default_rng(self.seed)
        logger.info(f"Using seed {self.seed}")

        # find out the number of sims
        self.nsims = settings.get("nsims", 100)
        self.narray = settings.get("narray", 1)
        self.total_sims = self.nsims * self.npol * self.narray

        self.npatches = settings.get("npatches", 10)
        self.total_patches = self.npatches * self.total_sims
        # self.npix = hp.nside2npix(self.nside)

        # some important derived parameters
        # most of these are not really used, delete?
        self.nell = self.lmax + 1
        self.nelem = hp.Alm.getsize(self.lmax)
        self.ells = np.arange(self.nell)

        if self.settings.get("double_precision", False):
            self.r_dtype = np.float64
            self.c_dtype = np.complex128
            self.precision = "double"
        else:
            self.r_dtype = np.float32
            self.c_dtype = np.complex64
            self.precision = "single"

        # get the beam and noise
        self.noise = settings.get("noise", True)
        self.beam_width = (settings.get("beam_width", 0) * u.arcmin).to_value(u.radian)
        noise_scale_tt = settings.get("noise_scale_tt", 1) * u.arcmin
        self.noise_scale_tt = noise_scale_tt.to_value(u.radian)
        noise_scale_ee = settings.get("noise_scale_ee", 1) * u.arcmin
        self.noise_scale_ee = noise_scale_ee.to_value(u.radian)
        noise_scale_te = settings.get("noise_scale_te", 1) * u.arcmin
        self.noise_scale_te = noise_scale_te.to_value(u.radian)
        self.noise_ell, self.beam_ell = self.get_noise_beam()

        # get the radii
        r_min = settings.get("r_min", 1)
        r_max = settings.get("r_max", 50000)
        self.radii, self.drs = self.get_radii(r_min, r_max)

        # get some info from SLURM
        self.sjob = os.getenv("SLURM_JOB_ID", 0)
        job_tasks = os.getenv("SLURM_ARRAY_TASK_COUNT")
        if job_tasks is not None:
            job_tasks = int(job_tasks)

            if job_tasks != self.narray:
                raise ValueError(
                    f"SLURM_ARRAY_TASK_COUNT {job_tasks} does not match narray value {self.narray}"
                )

            self.job_array_index = int(os.getenv("SLURM_ARRAY_TASK_ID", 0))
            ja_str = f"_{self.job_array_index}"

            if self.job_array_index == 1:
                self.is_main_job = True
            else:
                self.is_main_job = False

            logger.info(
                f"Running in SLURM job {self.sjob} with job array index {self.job_array_index} of {self.narray}"
            )
        else:
            logger.info(f"Running in SLURM job {self.sjob}")
            ja_str = ""

        # paths
        self.base_dir = settings.get("base_dir", "data")
        self.alm_dir = os.path.join(self.base_dir, settings.get("alm_dir", "alms"))
        self.plot_dir = os.path.join(self.base_dir, settings.get("plot_dir", "plots"))
        self.tb_dir = os.path.join(self.base_dir, settings.get("tb_dir", "tensorboard"))
        self.model_dir = os.path.join(
            self.base_dir, settings.get("model_dir", "models")
        )
        self.patch_dir = os.path.join(
            self.base_dir, settings.get("patch_dir", "patches")
        )

        # add info to the strings
        l_str = "l" if self.lensing else "ul"
        nn_str = "-nn" if not self.noise else ""

        # set up the file names
        pol_str = "".join(self.pols)

        self.base_name = settings.get(
            "base_name",
            f"l{self.lmax}_n{self.nside}_{l_str}{nn_str}_{pol_str}_{self.total_sims}",
        )
        logger.debug(f"Base name: {self.base_name}")

        # setup the file paths
        self.patch_str = f"{self.base_name}x{self.npatches}"
        self.patch_file_partial = os.path.join(
            self.patch_dir, f"{self.patch_str}{ja_str}.hdf5"
        )
        self.patch_file = os.path.join(self.patch_dir, f"{self.patch_str}.hdf5")

        self.alm_file_partial = os.path.join(
            self.alm_dir, f"{self.base_name}{ja_str}.hdf5"
        )
        self.alm_file = os.path.join(self.alm_dir, f"{self.base_name}.hdf5")

        # save a copy of the settings file if --save_settings is set
        if args.save_settings:
            dir = os.path.join("settings", "runs")
            os.makedirs(dir, exist_ok=True)

            file = os.path.join(dir, f"{self.sjob}_{self.base_name}.json")
            if not os.path.exists(file):
                logger.info(f"Saving run settings to file: {file}")
                with open(file, "w") as f:
                    json.dump(self.settings, f, indent=2)
            else:
                logger.warning(f"Settings file already exists: {file}, not overwriting")

        # print out the config object for the log
        logger.debug("Config Object:\n %s", vars(self))

    def get_noise_beam(self):
        if not self.noise:
            noise_ell = np.ones((self.nell), dtype=self.r_dtype) * 10**-16
            beam_ell = np.ones((self.nell), dtype=self.r_dtype)
            return noise_ell, beam_ell
        
        def remove_mono_dipole(alm):
            lmax = hp.Alm.getlmax(len(alm))
            alm[..., hp.Alm.getidx(lmax, 0, 0)] = 0.0
            alm[..., hp.Alm.getidx(lmax, 1, 0)] = 0.0
            alm[..., hp.Alm.getidx(lmax, 1, 1)] = 0.0
            return alm
        
        beam_ell_pre = hp.gauss_beam(self.beam_width, lmax=self.lmax, pol=True)
        beam_ell_pre = np.swapaxes(beam_ell_pre, 0, 1)

        noise_ell = []
        beam_ell = []
        if "T" in self.pols:
            beam_ell.append(beam_ell_pre[0])
            noise_map = np.random.normal(0, 1, hp.nside2npix(self.nside))
            noise_map = remove_mono_dipole(noise_map)
            noise = hp.anafast(noise_map, lmax=self.lmax, use_pixel_weights=True)
            noise_ell.append(noise * self.noise_scale_tt**2)
            
            # plt.figure()
            # plt.loglog(noise)
            # plt.xlabel('Multipole moment (l)')
            # plt.ylabel('Power')
            # plt.title('Noise power spectrum')
            # plt.show()

            # noise_map = hp.synalm(np.ones(nside * (nside + 1) // 2), lmax=nside - 1)
            # noise_cl = hp.anafast(noise_map)
            # n_ell = np.zeros((nside * (nside + 1) // 2, nside * (nside + 1) // 2))
            # for ell in range(nside * (nside + 1) // 2):
            #     for ell2 in range(nside * (nside + 1) // 2):
            #         n_ell[ell, ell2] = noise_cl[ell] * noise_cl[ell2] / (2 * ell + 1)

        # if "E" in self.pols:
        #     beam_ell.append(beam_ell_pre[1])
        #     noise = self.rng.normal(0, 1, self.nell) * self.noise_scale_ee**2
        #     noise_ell.append(noise)

        # if self.pols == ["T", "E"]:
        #     beam_ell.append(beam_ell_pre[3])
        #     noise = self.rng.normal(0, 1, self.nell) * self.noise_scale_te**2
        #     noise_ell.append(noise)

        return noise_ell, beam_ell

    def get_radii(self, r_min, r_max):
        # For the radii we follow Table 2. of Smith and Zaldarriaga which gives a greater density of points near reionization and recombination.
        # Spacing for all ranges but the last row are linear, with the last row having log spacing.
        # radii are in Mpc

        #    start,  stop, resolution
        ranges = [
            (0, 9500, 150),
            (9500, 11000, 300),
            (11000, 13800, 150),
            (13800, 14600, 400),
            (14600, 16000, 100),
            (16000, 50000, 100),
        ]
        radii = []

        for r in ranges:
            start = max(r_min, r[0])
            end = min(r_max, r[1])

            if start > end:
                continue

            if r == ranges[-1]:  # For the last range, use logspace
                temp_radii = np.logspace(np.log10(start), np.log10(end), num=r[2])
            else:
                temp_radii = np.linspace(start, end, num=r[2], endpoint=False)

            radii.extend(temp_radii)

        radii = np.array([r for r in radii if r_min <= r < r_max])
        drs = np.diff(radii) / 2.0
        return radii, drs
